src/transformers/models/zamba2/modular_zamba2.py

```diff
@@ -0,0 +1,1760 @@
+    Finally, this attention layer contributes to tied transformer blocks aimed to increasing compute without increasing model size. Because this
+    layer is tied, un-tied LoRA modules are added to the q, k, v projectors to increase expressivity with a small memory overhead.
+    """
+
+    def __init__(
+        self,
+        config: Zamba2Config,
+        layer_idx: Optional[int] = None,
+        num_fwd_mem_blocks: int = None,
+        block_id: int = None,
+    ):
+        super().__init__(config, layer_idx)
+        self.num_fwd_mem_blocks = num_fwd_mem_blocks
+        self.rope_theta = config.rope_theta
+        self.layer_block_map = layer_type_list(config)
+        self.block_id = block_id
+        self.is_causal = True
+
+        if config.use_shared_attention_lora:
+            self.linear_q_lora_A_list = nn.ModuleList([])
+            self.linear_q_lora_B_list = nn.ModuleList([])
+            self.linear_k_lora_A_list = nn.ModuleList([])
+            self.linear_k_lora_B_list = nn.ModuleList([])
+            self.linear_v_lora_A_list = nn.ModuleList([])
+            self.linear_v_lora_B_list = nn.ModuleList([])
+
+            for i in range(self.num_fwd_mem_blocks):
+                if i % config.num_mem_blocks == block_id:
+                    linear_q_lora_A = nn.Linear(self.attention_hidden_size, self.config.lora_rank, bias=False)
+                    linear_q_lora_B = nn.Linear(self.config.lora_rank, self.attention_hidden_size, bias=False)
+                    linear_k_lora_A = nn.Linear(self.attention_hidden_size, self.config.lora_rank, bias=False)
+                    linear_k_lora_B = nn.Linear(self.config.lora_rank, self.attention_hidden_size, bias=False)
+                    linear_v_lora_A = nn.Linear(self.attention_hidden_size, self.config.lora_rank, bias=False)
+                    linear_v_lora_B = nn.Linear(self.config.lora_rank, self.attention_hidden_size, bias=False)
+                else:
+                    linear_q_lora_A = nn.Identity()
+                    linear_q_lora_B = nn.Identity()
+                    linear_k_lora_A = nn.Identity()
+                    linear_k_lora_B = nn.Identity()
+                    linear_v_lora_A = nn.Identity()
+                    linear_v_lora_B = nn.Identity()
+                self.linear_q_lora_A_list.append(linear_q_lora_A)
+                self.linear_q_lora_B_list.append(linear_q_lora_B)
+                self.linear_k_lora_A_list.append(linear_k_lora_A)
+                self.linear_k_lora_B_list.append(linear_k_lora_B)
+                self.linear_v_lora_A_list.append(linear_v_lora_A)
+                self.linear_v_lora_B_list.append(linear_v_lora_B)
+
+        if config.use_mem_rope:
+            self.rotary_emb = Zamba2RotaryEmbedding(
+                config,
+                self.head_dim,
+                max_position_embeddings=config.max_position_embeddings,
+                base=self.rope_theta,
+            )
```

Comment:
I moved rope to model init ([commit](https://github.com/huggingface/transformers/pull/34517/commits/8079ae035614cdf662af92093c3a84887639821a)). 

As a remark, I overrode the forward pass of `ZambaModel` in `Zamba2Model`, since Zamba doesn't have rope. Similarly, I overrode the forward pass of `ZambaMambaDecoderLayer` in `Zamba2MambaDecoderLayer` because the former doesn't take `position_embeddings` as argument (and it wouldn't make sense to add it since there is no rope in Zamba).